lib/pbench/cli/agent/commands/tools/list.py

```diff
@@ -54,76 +57,68 @@ def execute(self) -> int:
-                        tool_info[group][host][tool] = (
-                            sorted((path / tool).read_text().rstrip("\n").split("\n"))
-                            if opts
-                            else ""
-                        )
-
-            if tool_info:
-                found = self.print_results(tool_info, self.context.with_option)
-                if not found:
-                    msg = "No tools found"
-                    if self.context.group:
-                        msg += f' in group "{self.context.group[0]}"'
-                    self.logger.warn(msg)
-            else:
-                self.logger.warn("No tool groups found")
-
-            return 0
+        for group in groups:
+            tool_info[group] = {}
+            try:
+                tg_dir = self.gen_tools_group_dir(group)
+            except BadToolGroup:
+                self.logger.error("Bad tool group: %s", group)
+                return 1
 
-        else:
-            # List the groups which include this tool
-            tool = self.context.name
-            found = False
-            for group in groups:
-                try:
-                    tg_dir = self.gen_tools_group_dir(group)
-                except BadToolGroup:
-                    self.logger.error("Bad tool group: %s", group)
-                    return 1
-
-                tool_info[group] = {}
-                for path in tg_dir.iterdir():
-                    # skip files like __label__ and __trigger__
-                    if not path.is_dir():
-                        continue
-
-                    host = path.name
-                    tool_info[group][host] = {}
-                    # Check to see if the tool is in any of the hosts.
-                    if tool in self.tools(path):
-                        tool_info[group][host][tool] = (
-                            sorted((path / tool).read_text().rstrip("\n").split("\n"))
-                            if opts
-                            else ""
-                        )
+            for path in tg_dir.iterdir():
+                # skip __trigger__ if present
+                if not path.is_dir():
+                    continue
+
+                host = path.name
+                tool_info[group][host] = {}
+
+                if toolname:
+                    tool_info[group][host]["label"] = None
+                else:
+                    label = path / "__label__"
+                    v = label.read_text().rstrip("\n") if label.exists() else None
+                    tool_info[group][host]["label"] = v
+
+                toolsdict = tool_info[group][host]["tools"] = {}
+                if toolname:
+                    # Check if the tool is in any of the hosts.
+                    if toolname in self.tools(path):
+                        v = (path / toolname).read_text().rstrip("\n").split("\n")
+                        toolsdict[toolname] = sorted(v) if opts else ""
```

Comment:
While I'm pointing this out, I'm not sure I want it to change...

Both here and on line 97 below, you're doing the path-join/read/strip/split, which isn't "free", and then completely ignoring the result `if not opt`, which is a value computed way up top. It'd be a bit more efficient not to compute the value we already know we're not going to use.

On the other hand, this adds an ugly `if` to already way indented blocks of code, and I'm uncertain whether I like that tradeoff ...